### PR TITLE
ENH: Add ML training and inference

### PR DESCRIPTION
**Pull request by nidhinthomas-ai**
## Summary

This MR introduces a compact ML toolkit for phase diagram construction including and model training, inference, explainability and plotting:

* **Training**: an RF+XGBoost soft‑voting pipeline with simple pre‑processing, class‑imbalance handling, and a lightweight grid search over a validation set.
* **Inference**: robust prediction on new CSVs, feature alignment against a saved model bundle, and CSV export of probabilities and labels.
* **Explainability**: SHAP, EBM, and LIME importances with a consensus view and publication‑quality figures.
* **Utilities**: helpers for plotting/analysis, including tools for GMM‑based decision regions & phase‑boundary extraction in composition space.

---

## What’s included

```
neat_ml/
├── model/
│   ├── train.py                # training pipeline, tuning, ROC plotting, model bundling
│   ├── inference.py            # load bundle, align features, predict, save CSV
│   └── feature_importance.py   # SHAP / EBM / LIME + consensus and plots
└── utils/
    └── utils.py                # plotting helpers, GMM wrappers, boundaries, styling
```

---

## Design highlights

* **Ensemble learner**: `VotingClassifier` over `RandomForestClassifier` and `XGBClassifier` with soft voting.
* **Pre‑processing**:
  * All features coerced to numeric, **columns that are entirely NaN are dropped**.
  * Median imputation (performed in `preprocess()` and again inside the pipeline for robustness when new features are missing at inference).
  * Standardization via `StandardScaler`.
* **Imbalance handling**: uses `scale_pos_weight = (#neg / #pos)` for XGBoost, computed on the training (and refit on train+val) target.
* **Hyperparameter search**: exhaustive product over a small grid (72 combos; see below) scored by **validation ROC‑AUC**; best model is then refit on combined train+val.
* **Reproducibility**: `RANDOM_STATE=42` across learners.
* **Model artifact**: single `joblib` bundle containing the fitted pipeline, feature list, validation metrics, and best parameters.

---

### `neat_ml/model/train.py`

**Public API**

* `preprocess(df, target, exclude=None) -> (X, y)`
  * Drops rows with missing target, coerces numerics, drops all‑NaN columns, median‑imputes remaining NaNs.
* `train_with_validation(X_train, y_train, X_val, y_val)`
  * Grid‑searches the ensemble, picks best by **ROC‑AUC**, reports **PR‑AUC** for the chosen model, then **refits** on `train+val`.
  * **Returns**: `(final_model_pipeline, metrics_dict, best_params_dict, val_proba_best)`.
* `plot_roc(y_true, y_prob, out_png, label="Validation")`
  * Saves a 4x4 in ROC figure (PNG).
* `save_model_bundle(model, features, metrics, best_params, path)`
  * Serializes `{ "model": Pipeline, "features": List[str], "metrics": Dict[str, float], "best_params": Dict[str, Any] }`.

**Model pipeline**

```text
[ SimpleImputer(median) ] → [ StandardScaler ] → [
  VotingClassifier(
    estimators = [
      ("rf",  RandomForestClassifier(class_weight="balanced")),
      ("xgb", XGBClassifier(objective="binary:logistic", scale_pos_weight=...))
    ],
    voting="soft"
  )
]
```

**Hyperparameter grid (72 total)**

* RF: `n_estimators=[500]`, `max_depth=[None]`
* XGB: `n_estimators=[10,20,50,100,200,400]` × `learning_rate=[0.05,0.1]` × `max_depth=[3,4,5,8,10,20]`

> Note: RF/XGB have defaults in `_build_pipeline` that are overridden via `.set_params(**grid_params)` during search.

---

### `neat_ml/model/inference.py`

**Public API**

* `save_predictions(df, y_prob, out_csv)`
  * Appends:
    * `Pred_Prob` (positive‑class probability)
    * `Pred_Label` (`1` if `>=0.5`, else `0`)
  * Writes CSV.
* `run_inference(model_in, data_csv, target: Optional[str], exclude_cols: list[str], pred_csv)`
  * Loads the `joblib` bundle, reads `data_csv`, runs `preprocess` (with optional `target` to enable label alignment but **no evaluation/ROC** is performed here), **adds any missing training features** as all‑NaN columns, reorders to match the saved feature list, and computes probabilities with the bundled pipeline.
  * Calls `save_predictions(...)`.

### `neat_ml/model/feature_importance.py`

**Public API**

* `compare_methods(model, X, y, out_dir, top=20)`
  * Runs:
    * **SHAP (permutation)**: saves `shap_summary.png`; returns mean |SHAP| importances.
    * **EBM (ExplainableBoostingClassifier)**: saves `ebm_importance.png` & `ebm_importance.csv`.
    * **LIME** (aggregated over sampled rows): returns mean |weight| importances.
  * Merges importances into `feature_importance_comparison.csv` and `feature_importance_comparison.png`, ranked by **mean rank** across methods.
  * Builds a **consensus** (majority vote across methods for top‑k) and saves `feat_imp_consensus.png`.
* `feature_importance_consensus(pos_class_feat_imps, feature_names, top_feat_count)`
  * Majority‑vote ranking across importance vectors; returns `(ranked_names, ranked_counts, num_models)`.
* `plot_feat_import_consensus(ranked_names, ranked_counts, num_models, top_feat_count, out_dir)`
  * Horizontal bar chart of **% occurrence** in top‑k per method.
* `get_k_best_scores(X, y, k, metrics)`
  * Convenience helper to collect `SelectKBest` scores (raw vectors) for various scoring functions.

**Dependencies**

* Requires `shap`, `interpret` (glassbox), and `lime`. Currently imported at module import time (i.e., behave as **hard dependencies**). See TODOs for lazy import to make them optional.

---

### `neat_ml/utils/utils.py`

**Public API & behavior**

* `_axis_ranges(df1, df2, x_col, y_col, pad=2) -> ([0,x_max],[0,y_max])` Shared axis ranges across two frames (integer bounds).
* **GMM utilities**
  * `GMMWrapper(gmm, x_comp, x_features)` Uses a `GaussianMixture` trained on 1D **phase features** to classify any (x,y) **composition** point by nearest neighbor mapping back to the observed feature domain.
  * `_standardise_labels(cluster_labels, x_comp)` Deterministic remap of raw GMM labels to the convention:
    * `0 = Two Phase (triangle‑up, aquamarine)`
    * `1 = Single Phase (square, light‑steel‑blue)`
  * `extract_boundary_from_contour(z, xs, ys, level=0.5)` From a grid `z`, returns the longest (x,y) contour at threshold `level` (e.g., phase boundary).
  * `plot_gmm_decision_regions(df, x_col, y_col, phase_col, ax, xrange, yrange, ..., plot_regions=True, ...) -> (gmm, std_labels, boundary_points)`
    * Fits GMM on `phase_col`, projects decisions onto composition grid (xrange×yrange), optionally **fills regions** and draws the **decision boundary**; returns model, standardized labels for observed points, and extracted boundary coordinates.
  * `plot_gmm_composition_phase(df, x_col, y_col, phase_col, ax, point_cmap=None)`
    * Fits GMM on `phase_col`, scatters (x,y) points with standardized phase labels and legend.

**Styling helpers**

* `_set_axis_style(ax, xrange, yrange)` Equal aspect, integer tick formatting, thicker spines/ticks.

---

### 1) Train -\> validate -\> bundle

```python
import pandas as pd
from sklearn.model_selection import train_test_split
from neat_ml.model.train import preprocess, train_with_validation, plot_roc, save_model_bundle

df = pd.read_csv("train.csv")
X, y = preprocess(df, target="label", exclude=["id"])  # binary labels 0/1 expected

X_tr, X_val, y_tr, y_val = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

model, metrics, best_params, val_proba = train_with_validation(X_tr, y_tr, X_val, y_val)

print(metrics)       # {'val_roc_auc': ..., 'val_pr_auc': ...}
print(best_params)   # chosen RF/XGB params

plot_roc(y_val, val_proba, out_png="artifacts/val_roc.png")

save_model_bundle(
    model=model,
    features=X.columns.tolist(),
    metrics=metrics,
    best_params=best_params,
    path="artifacts/model_bundle.joblib",
)
```

### 2) Run inference on a new CSV

```python
from pathlib import Path
from neat_ml.model.inference import run_inference

run_inference(
    model_in=Path("artifacts/model_bundle.joblib"),
    data_csv=Path("inference.csv"),
    target=None,                 # or "label" if present; used only for alignment
    exclude_cols=["id"],
    pred_csv=Path("predictions.csv"),
)
# predictions.csv includes: ..., Pred_Prob, Pred_Label
```

### 3) Explainability & consensus

```python
import pandas as pd, joblib
from neat_ml.model.train import preprocess
from neat_ml.model.feature_importance import compare_methods

bundle = joblib.load("path_to_model_bundle.joblib")
model = bundle["model"]

df = pd.read_csv("train.csv")
X, y = preprocess(df, target="label", exclude=["id"])

compare_methods(model, X, y, out_dir=Path("explain/"), top=20)
# Produces: shap_summary.png, ebm_importance.png/.csv,
#           feature_importance_comparison.png/.csv, feat_imp_consensus.png
```

## MR checklist

* [x] New modules and functions documented above
* [x] Artifacts written to disk (bundle, plots, CSVs)
* [x] Basic logging via `print(...)`
* [x] Unit tests
* [x] ruff check
* [x] mypy testing

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,191 @@
+name: CI
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:   # run for all PRs (parity with GitLab MR rule)
+    types: [opened, synchronize, reopened, ready_for_review]
```
Comment (tylerjereddy) [inline .github/workflows/ci.yml]: Purge this out--seems unnecessary. In general, this CI config file has far too much duplication/complexity, and merge conflicts need to be resolved against the more concise CI config I added in gh-9.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        run: |
+          set -euxo pipefail
+          python -m pip install -U pip setuptools wheel
+          python -m pip install \
+            mypy ruff pandas-stubs \
+            "numpy>=1.26.3,<=2.1.3" matplotlib xgboost pandas scikit-learn \
+            openpyxl jinja2 python-ternary pytest shap interpret \
+            types-tqdm types-Pillow scikit-image lightgbm lime \
+            opencv-python-headless==4.10.0.84 \
+            pyyaml types-pyyaml torch
```
Comment (tylerjereddy) [inline .github/workflows/ci.yml]: this is verbose and gets repeated below--I think we should just list the dependencies in `pyproject.toml` and use the standard approach of installing them with `pip` from that file...

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+          cache: pip
+          cache-dependency-path: |
+            pyproject.toml
+            setup.cfg
+            **/requirements*.txt
+            neat_ml/sam2/pyproject.toml
+            neat_ml/sam2/setup.cfg
```
Comment (tylerjereddy) [inline .github/workflows/ci.yml]: get rid of caching for this small team/project

File: neat_ml/analysis/data_analysis.py [hunk 4]
```diff
@@ -0,0 +1,747 @@
...
+                        dist = float(dists[i, nn_idx])
+                        graph.add_edge(i, j, distance=dist)
+        elif method == "delaunay" and n_nodes >= 3:
+            tri = Delaunay(points)
+            for simplex in tri.simplices:
+                for i, j in zip(simplex, np.roll(simplex, -1)):
+                    if not graph.has_edge(i, j):
+                        dist = float(np.linalg.norm(points[i] - points[j]))
+                        graph.add_edge(int(i), int(j), distance=dist)
+    except Exception as exc:
```
Comment (tylerjereddy) [inline neat_ml/analysis/data_analysis.py]: Let's try to avoid capturing the generic `Exception` class for the usual reasons

File: neat_ml/bubblesam/bubblesam.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+            (min_col, min_row),
+            max_col - min_col,
+            max_row - min_row,
+            linewidth=1,
+            edgecolor='green',
+            facecolor='none'
+        )
+        ax.add_patch(rect)
+    ax.axis("off")
+    plt.savefig(output_path, bbox_inches='tight')
```
Comment (tylerjereddy) [inline neat_ml/bubblesam/bubblesam.py]: Here and elsewhere--let's try to use `fig`, `ax` objects instead of the "global" `plt` approach, for the usual reasons.

File: neat_ml/model/train.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        - A dictionary of the best hyperparameters
+          found during the grid search.
+        - The predicted probabilities for the positive
+          class on the validation set.
+    """
+    spw = _scale_pos_weight(y_train)
+    print(f"[INFO] scale_pos_weight={spw:.3f}  |  train neg/pos={np.bincount(y_train)}")
+
+    param_names = list(_PARAM_GRID.keys())
+    param_values = list(_PARAM_GRID.values())
```
Comment (tylerjereddy) [inline neat_ml/model/train.py]: let's avoid using "module globals" like `_PARAM_GRID`, instead generally passing variables in through functions arguments to avoid spaghetti logic...

File: neat_ml/model/train.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+        print(f"[GRID] params={params}  |  val ROC-AUC={auc:.4f}")
+
+        if auc > best_auc:
+            best_auc = auc
+            best_params = params
+            best_model = model
+            val_proba_best = val_proba
+
+    assert best_model is not None, "No model was fitted!"
+    assert val_proba_best is not None, "Validation probabilities were not calculated!"
```
Comment (tylerjereddy) [inline neat_ml/model/train.py]: avoid plain `assert` in production code, for the usual reasons

File: neat_ml/model/train.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    fpr, tpr, _ = roc_curve(y_true, y_prob)
+    aucv = roc_auc_score(y_true, y_prob)
+    plt.figure(figsize=(4, 4))
+    plt.plot(fpr, tpr, lw=2, label=f"{label} AUC={aucv:.3f}")
+    plt.plot([0, 1], [0, 1], "--", lw=1, color="grey")
+    plt.xlabel("False-Positive Rate")
+    plt.ylabel("True-Positive Rate")
+    plt.legend()
+    plt.tight_layout()
+    plt.savefig(out_png, dpi=300)
```
Comment (tylerjereddy) [inline neat_ml/model/train.py]: Let's avoid the "global"/MATLAB-like `plt` usage, instead favoring the object-oriented `fig`, `ax` approach..

File: neat_ml/model/train.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+    plt.close()
+    print(f"[INFO] ROC curve saved -> {out_png}")
```
Comment (tylerjereddy) [inline neat_ml/model/train.py]: `logging` not `print()` for production output

**Comment by tylerjereddy** [comment]
@adamwitmer paper revisions are due in a few weeks max, and it could very well take me a few weeks to review this even if it were ready for review *now*, so definitely way past due to have this ready for review given December 2025 deadline.

**Comment by tylerjereddy** [comment]
superseded by the merging of gh-28, closing
